### PR TITLE
KAFKA-7944: Improve Suppress test coverage

### PR DESCRIPTION
**Pull request by vvcephei**
* add a normal windowed suppress with short windows and a short grace
  period
* improve the smoke test so that it actually verifies the intended
  conditions

See https://issues.apache.org/jira/browse/KAFKA-7944

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/integration/SmokeTestDriverIntegrationTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/SmokeTestDriverIntegrationTest.java:82]: More stable and future-proof test setup code. I needed this because I added a second test here, which I've since removed.

Leaving this change, though, so we won't have to waste time debugging again next time we try to add a test.

File: streams/src/test/java/org/apache/kafka/streams/integration/SuppressionDurabilityIntegrationTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/SuppressionDurabilityIntegrationTest.java]: unused!

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:119]: I added this while debugging the smoke test. Printing the time allows us to correlate events across test nodes and with the test logs.

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:170]: Added a "small window sum" stream to the topology, which does a "normal" windowed computation. The exact data production timing is non-deterministic, so we can't verify the results, but we can verify that there is exactly one result per windowed key in the suppressed stream

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java:63]: Apparently this is unnecessary now?

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java:84]: little bit of cleanup.

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: Ability to slow down data production so that the system tests can have time to bounce nodes *while* processing instead of *after* processing.

I happened to examine the smoke-test system test logs, and realized that we were fully processing the results before the first bounce. This totally circumvents the purpose of the test.

To get the prior behavior, you can just pass `Duration.ZERO`. The computation below will yield a record pause time of `0`, and then the sleep max will produce the prior value of `2ms`.

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java:447]: Decomponsed `verifySuppressed` and `verify`, so that we can just verify that `sws-suppressed` is actually suppressed, without verifying the results (which we would have to compute from first principles)

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java:520]: relaxing type constraints

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java:531]: the second cast is unnecessary to prevent overflow

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java]: since this is just going to run until stopped, there's no need (or any sensible value) for delay.

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java:82]: added the ability to run the smoke test app with EOS

File: tests/kafkatest/services/streams.py
Comment (vvcephei) [inline tests/kafkatest/services/streams.py:358]: Let me know if there's a better way to do this...

File: tests/kafkatest/tests/streams/streams_bounce_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_bounce_test.py:23]: This is redundant now, since the smoke_test system test does crashes in addition to graceful stops now.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:51]: Added this so we can test with/without EOS and with graceful/hard shutdown

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:61]: I went down to 3 processors and changed up the test scenario a little.

Instead of starting two nodes concurrently at the start, we just start one so that we can be sure it will process some data before we kill it. The same logic applies to the second round of processing and kill.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:70]: Wait for Streams to start first (so we don't generate all the data before streams is even started)

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:72]: start generating data. This will run for 30 seconds before verifying. We have verifications in place to make sure we don't finish processing before we do both node bounces.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:77]: Wait until we actually process some data (it's pointless if we bounce before we process anything)

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:80]: Here's where we check that we're not already done processing (it would be pointless to bounce after processing is complete).

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:82]: either stop the node gracefully or crash it (`kill -9`), depending on the parameter.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:88]: Again, wait for the node to start. This isn't strictly necessary, since we're just going to wait for "processed" next, but I thought it made the tests more readable and debuggable. For example, when you get a test failure, you'll know whether the node failed to start or whether it started, but didn't process anything.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:96]: Again, before bouncing the node, make sure the verification isn't already done, which would be pointless.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:115]: The test is already over. We can stop the app gracefully.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:118]: If we crash without EOS, we might process some duplicates (this is the whole point of EOS), so Streams is operating properly if it *either* does exactly the right thing or processes too many records.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [inline tests/kafkatest/tests/streams/streams_smoke_test.py:120]: In all other cases, we expect to get exactly the right result.

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @mjsax @bbejeck @ableegoldman ,

In response to a mailing list message that they still saw a problem with Suppress after the fix, I bumped up the priority of KAFKA-7944. (this was similar to the reporter's topology)

I also took the time to double check we weren't getting a false negative (test passing without the intended verification), and learned that actually we could get false negatives for a number of reasons:
* we could process the entire data set before the first bounce.
* sometimes, the node that we bounce hadn't actually processed anything (it was all assigned to the other node).

I also wanted to verify the smoke test app with EOS.

Since I took the time to refactor the smoke_test system test to achieve these goals, and there was a large overlap with the bounce test, I also unified the tests by adding the `crash` parameter to the smoke test.

I requested a review from you all so that we can get as much scrutiny on this system test as possible. Trying not to create a new flaky test here...

**Comment by vvcephei** [comment]
System tests:
* just smoke test:
  * https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2387/
  * passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-03-06--001.1551897420--vvcephei--KAFKA-7944-suppress-window-test--efb4407/report.html
* streams tests:
  * https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2388/

**Comment by bbejeck** [comment]
\cc @ableegoldman for review

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: good call!

File: tests/kafkatest/services/streams.py
Comment (bbejeck) [reply tests/kafkatest/services/streams.py:358]: I thought there were already existing`EOS` tests that use the `StreamsSmokeTest`?

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (bbejeck) [inline tests/kafkatest/tests/streams/streams_smoke_test.py]: Don't we want to at least `processor3` starts successfully here?

**Review by bbejeck — APPROVED** [review]
Thanks for the updated test @vvcephei this LGTM (assuming system tests pass) just left a couple of minor comments.

**Comment by guozhangwang** [comment]
`org.apache.kafka.streams.integration.SuppressionDurabilityIntegrationTest.initializationError` consistently failing.

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestClient.java:119]: Nice find!

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: Hmm this seems an IDE specific as well? https://github.com/rtyley/intellij-community-set-committer-to-author/blob/master/plugins/InspectionGadgets/src/inspectionDescriptions/DynamicRegexReplaceableByCompiledPattern.html

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: hey @vvcephei I'm wondering if we can simplify the code further: with `timeToSpend` now can we just remove autoTerminate?

1) SmokeTestDriverIntegrationTest.java: we are sending 1000 * 10 = 10K records only, so likely that will stop even before the second instance can be started, similar to the issue you observed in system test. I'd suggest we just set `timeToSpend` to 10 seconds so we have some enough time to start up to 10 / 1 (sleep time) = 10 instances.

2) As for StreamsSmokeTest itself, we only disableAutoTerminate in three of StreamsUpgradeTest cases, since we do not need to verify the number of records sent / received but only check upgrade completed. We can also use 30 seconds (not sure if it is sufficient, but we can get on average how much time those three tests will take from our nightlies), and even if the test completes before that it will call `driver.stop` to force-stop the driver anyways.

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: Of course we can then remove the whole `disableAutoTerminate` thing from StreamsSmokeTest.

File: tests/kafkatest/tests/streams/streams_bounce_test.py
Comment (guozhangwang) [reply tests/kafkatest/tests/streams/streams_bounce_test.py:23]: +1

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsSmokeTest.java:82]: Can we use `StreamsEosTest` instead? See my other comment below.

File: tests/kafkatest/services/streams.py
Comment (guozhangwang) [reply tests/kafkatest/services/streams.py:358]: To my other comment above: what's the difference of running `StreamsSmokeTestBaseService#process-eos` v.s. `StreamsEosTestBaseService#process`? The former uses StreamsSmokeTest while the latter use StreamsEosTest client. Can we just consolidate them?

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (guozhangwang) [reply tests/kafkatest/tests/streams/streams_smoke_test.py:118]: sgtm.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (guozhangwang) [reply tests/kafkatest/tests/streams/streams_smoke_test.py:88]: Not a comment: I actually think it is indeed necessary to avoid flakiness, remember @bbejeck once talked about it in an older PR?

**Comment by bbejeck** [comment]
Both Java 8 and Java 11 failed but build results already cleaned up.

retest this please

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: Hmm, this is an interesting idea. I'll look into it.

That's a very good point about the integration test, I'll look at that as well.

File: tests/kafkatest/services/streams.py
Comment (vvcephei) [reply tests/kafkatest/services/streams.py:358]: Shameful admission: I thought about this, but just didn't take the time to check.

The bounce test was obviously duplicated, but the eos test is more complex. I knew I'd have to do a case-by-case check to make sure I didn't remove coverage. I'll do my homework now...

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [reply tests/kafkatest/tests/streams/streams_smoke_test.py:88]: I think that was a different situation, like maybe we stopped and then re-started a node? In that case, you need to grab a new monitor after you stop and before you re-start, to be sure your grep won't match the "started" message from the first time it was running.

Since we're querying the same monitor here in both blocks, we're grepping over the same range of the file. The monitor pins the start-point for grepping when you create it. For example, since we create this monitor before we start Streams, it uses "byte 1" as the starting points for all greps for both `REBALANCING -> RUNNING` and `processed`.

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [reply tests/kafkatest/tests/streams/streams_smoke_test.py]: We could wait on that, it might make the test more readable. But it's not strictly necessary, since if it doesn't start, then we won't finish processing.

I'll add it for readability.

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: Ah, oops. I'll remove it.

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]: Ok, After trying to simplify the code, I found that it was actually better to split the two use cases, so that's what I did. It lets us ditch a whole bunch of bookkeeping when we just want to generate until the process is killed, and it also lets us skip a bunch of mind-bending boolean logic when we do want to just generate a fixed amount of data.

I also added a duration of 20 seconds to the integration test, and it didn't increase the run-time of the test, but now we're getting the same guarantee that we get with the system tests now.

File: tests/kafkatest/services/streams.py
Comment (vvcephei) [reply tests/kafkatest/services/streams.py:358]: Ok, so the smoke test and the eos test are similar, but not identical.

The smoke test application has more features in it, though. So, we have more feature coverage under eos when we test with the smoke test.

The eos test evaluates two topologies one with no repartition, and one with a repartition. The smoke test topology contains repartitions, so it only tests _with_ repartition. I think that it should be sufficient to test only _with_ repartition.

The eos test verification specifically checks that "all transactions finished" (`org.apache.kafka.streams.tests.EosTestDriver#verifyAllTransactionFinished`). I'm not clear on exactly what we're looking for here. It looks like we create a transactional producer and send a record to each partition and then expect to get all those records back, without seeing any other records. But I'm not sure why we're doing this. If we want to drop the eos test, then we might need to add this to the smoke test verification.

WDYT?

File: tests/kafkatest/tests/streams/streams_smoke_test.py
Comment (vvcephei) [reply tests/kafkatest/tests/streams/streams_smoke_test.py]: Done, it occurred to me during the change that it also has the side benefit of making sure that the third processor does process some events.

**Comment by vvcephei** [comment]
Jenkins threw some crazy exception during the system tests. Re-running:
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2392

System tests passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-03-07--001.1552001126--vvcephei--KAFKA-7944-suppress-window-test--48dabcd/report.html

**Comment by vvcephei** [comment]
Reported unrelated failure: https://issues.apache.org/jira/browse/KAFKA-7965

Retest this, please.

**Comment by vvcephei** [comment]
test results are unavailable.

Retest this, please.

File: streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/tests/SmokeTestDriver.java]:
```suggestion
```

File: tests/kafkatest/services/streams.py
Comment (guozhangwang) [reply tests/kafkatest/services/streams.py:358]: `verifyAllTransactionFinished` aimed to avoid a situation that some dangling txn is open forever, without committing or aborting. Because the consumer needs to guarantee offset ordering when returning data, with read-committed they will also be blocked on those open txn's data forever (this usually indicates a broker-side issue, not streams though, but still).

I think we should still retain this check if we want to merge in the EosTest to SmokeTest, but also if you do not want to drag on it we could keep it as a separate ticket and merge this as-is. Your call :)

**Comment by guozhangwang** [comment]
LGTM to merge as-is (replied to @vvcephei 's question about eos test), leave it to @bbejeck for merging

File: tests/kafkatest/services/streams.py
Comment (vvcephei) [reply tests/kafkatest/services/streams.py:358]: Ok, thanks for the explanation. Let's merge this, and I'll create a ticket to follow up by including this check and removing the eos test.

That'll help review burden as well, I think.

**Comment by vvcephei** [comment]
Thanks for the response, @guozhangwang . I've created https://issues.apache.org/jira/browse/KAFKA-8080 to follow up.

**Comment by vvcephei** [comment]
test results are unavailable.

Retest this, please.

**Comment by vvcephei** [comment]
Thanks, @guozhangwang !

**Pull request merged**
